tests/unit/testsuite/Bolt/Boltpay/Controller/Traits/WebHookTraitTest.php

```diff
@@ -0,0 +1,441 @@
+    }
+
+    /**
+     * @test
+     * Pre-dispatch method with validating signature
+     *
+     * @covers Bolt_Boltpay_Controller_Traits_WebHookTrait::verifyBoltSignature
+     * @covers Bolt_Boltpay_Controller_Traits_WebHookTrait::preDispatch
+     * @dataProvider payloadProvider
+     * @expectedException Exception
+     * @expectedExceptionMessage Avoid calling parent::preDispatch
+     *
+     * @param string $payload Webhook payload in JSON format
+     * @throws ReflectionException from TestHelper if a specified object, class or property does not exist.
+     */
+    public function preDispatch_withRequestSignatureValidationEnabled_shouldTryToVerifyHook($payload)
+    {
+        $this->response->expects($this->once())->method('clearAllHeaders')->willReturnSelf();
+        $this->response->expects($this->once())->method('clearBody')->willReturnSelf();
+        $this->helperMock->expects($this->once())->method('setResponseContextHeaders');
+        $this->response->expects($this->once())->method('setHeader')->with('Content-type', 'application/json', true);
+        $this->layout->expects($this->once())->method('setDirectOutput')->with(true);
+        TestHelper::setNonPublicProperty(
+            $this->currentMock,
+            'requestMustBeSigned',
+            true
+        );
+
+        $_SERVER['HTTP_X_BOLT_HMAC_SHA256'] = self::TEST_HMAC;
+        //populate php://input stream
+        Bolt_Boltpay_StreamHelper::setData($payload);
+
+        $this->helperMock->expects($this->once())->method('verify_hook')->with($payload, self::TEST_HMAC)
+            ->willThrowException(new Exception('Avoid calling parent::preDispatch'));
+
+        // check to disable output buffering since ob_start() is called inside preDispatch
+        if (ob_get_level() > 0) {
+            ob_end_clean();
+        }
+
+        Bolt_Boltpay_StreamHelper::register();
+        $this->currentMock->preDispatch();
+    }
+
+    /**
+     * @test
+     * Getting request data from payload property
+     *
+     * @covers Bolt_Boltpay_Controller_Traits_WebHookTrait::getRequestData
+     * @dataProvider payloadProvider
+     *
+     * @param string $payload in JSON format
+     * @throws ReflectionException from TestHelper if a specified object, class or property does not exist.
+     */
+    public function getRequestData_withVariousPayload_shouldReturnPayloadAsObjectByDecodingJSON($payload)
+    {
+        TestHelper::setNonPublicProperty(
+            $this->currentMock,
+            'payload',
+            $payload
+        );
+        $result = TestHelper::callNonPublicFunction(
+            $this->currentMock,
+            'getRequestData'
+        );
+        $this->assertEquals(
+            json_decode($payload),
+            $result
+        );
+    }
+
+    /**
+     * Data provider for payload data in JSON format
+     *
+     * @return array containing payload data in JSON format
+     */
+    public function payloadProvider()
```

Comment:
Small nit - I think it's a little neater for organization to have all the dataProviders at the bottom. Or at least, not have [TEST] [Provider] [TEST].